extension-manifest-v3/src/store/options.js

```diff
@@ -75,40 +118,80 @@ const Options = {
             : options,
       });
 
-      // Send update message to another contexts (background page / panel / options)
-      try {
-        chrome.runtime.sendMessage({
-          action: UPDATE_OPTIONS_ACTION_NAME,
-          options,
-        });
-      } catch (e) {
-        console.error(
-          `Error while sending update options to other contexts: `,
-          e,
-        );
-      }
+      Promise.resolve(keys.includes('revision') ? null : sync(options, keys))
+        .then(() =>
+          // Send update message to another contexts (background page / panel / options)
+          chrome.runtime.sendMessage({
+            action: UPDATE_OPTIONS_ACTION_NAME,
+          }),
+        )
+        .catch(() => null);
+
       return options;
     },
-    observe: (_, options) => {
-      observers.forEach((fn) => fn(options));
+    observe: (_, options, prevOptions) => {
+      // TIP: First time or revision did not change (e.g. store.clear() was called)
+      if (!prevOptions || prevOptions.revision === options.revision) {
+        sync(options);
+      }
+
+      observers.forEach(async (fn) => {
+        try {
+          await fn(options, prevOptions);
+        } catch (e) {
+          console.error(`Error while observing options: `, e);
+        }
+      });
     },
   },
 };
 
 export default Options;
 
+async function sync(options, keys) {
+  if (options.terms && options.sync) {
+    const serverOptions = await getUserOptions();
+    const revision = options.revision;
+
+    if (serverOptions) {
+      if (serverOptions.revision > revision) {
+        const values = {};
+        for (const key of SYNC_OPTIONS) {
+          if (keys && keys.includes(key)) continue;
+          values[key] = hasOwnProperty.call(serverOptions, key)
+            ? serverOptions[key]
+            : options[key];
+        }
+
+        options = await store.set(Options, {
+          ...values,
+          revision: serverOptions.revision + (keys ? 0 : 1),
+        });
+      }
+
+      // TIP: Keys are only passed when the options are set
+      if (serverOptions.revision < revision || keys) {
+        await setUserOptions(
```

Comment:
With discussed protocol changes, the server will increase the version if it accepts the change. Will depend how the server implements it, but the client will see from the result whether it was accepted or not. If it was accepted, the client should also update its local version, since it would otherwise be behind the server and it would not be able to update till the next sync.

(One idea discussed is that we change the server to return 202 (ACCEPTED) if it updated the state. In that case, the version will be one higher than the client. And if it rejected, 409 (CONFLICT). In that case, the client should not update its local version. But it may be a good time to poll for the new changes in that situation.)

Reply:
Let's consider the situation when a client has local changes, but knows that its version is outdated. In other words, the user changed some other settings in the meantime. These other changes have also have propagated to other clients. 

What to do with the local changes now? In the end, it is a UI question, but these are the options that I can see:

1. Update to the latest version and discard the local changes (with or without visual feedback)
2. Stay on the local version (with or without visual feedback, something like "out of sync, syncing is now disabled). Optionally ask the user to resolved it (e.g. turn off sync forever, force a sync to the current version and discard local changes, or force-push the local changes and lose the updates on the other devices)
3. By default, force-push changes (in other words, ignore the conflict, with or without visual feedback). The drawback is that the other client will now lose the changes (in some sense, we would be again but to the same problem)
4. Resolve the conflict without user interaction and push the merged version as the new state.

For me personally, option 1 (without any visual feedback) would be OK. Optionally, having a warning that the client is not able to sync is nice. For instance, warn if the client has local changes, but fails to sync them to the server.

Staying on the old version is possible (option 2), but then you are stuck in that state and syncing will be disabled. The use can still change settings locally; but unless the client gets in sync again, these changes will not reach other clients. We could ask the user how to proceed (discard or force-push) to enable it again. On the plus side, it would give the user more control.

Always automatically force-pushing (option 3) creates more problems that it solves IMHO. It turns off optimistic locking and switches to the model as we have in MV2 Ghostery.

Option 4 might be the best user experience if it works. But it is not easy to implemented correctly. Though the extension could try to merge and if it fails, fall back to one of the other strategies.

(Personally, I'm leaning towards option 1 and would hope that it doesn't happen too often.)